### ISSUE TITLE
Using an OAuth implicit grant flow for non-login purposes logs the user out

### ISSUE THREAD

**Issue by elivox** [supabase/gotrue-js#527]
# Bug report

## Describe the bug

If a site using Supabase Auth uses an OAuth implicit grant flow for reasons _other_ than managing user accounts (e.g., fetching data from a third party website), then upon redirect, GoTrueClient will automatically detect the `#access_token` fragment, and attempt create a session using that token. 

This token, naturally, cannot be used for Supabase authentication. The resulting error causes any existing user session to be removed.

## To Reproduce

1. Go to supabase.com, and log in to your user account.
2. Append `#access_token=anystring` to the URL, and reload the page.
3. Observe that you have been logged out.

This can also be demonstrated in a fresh project such as https://github.com/supabase/supabase/tree/master/examples/user-management/nextjs-ts-user-management

Any OAuth grant flow that places an unrecognized `#access_token` in the URL will also log out the user, whether it was intended to or not.

## Expected behavior

The user remains logged in during OAuth flows unrelated to user account management.
Failing that, the app clearly communicates that
1. A user login process has been initiated due to URL contents.
2. The login process failed due to an invalid access token.

## System information

Reproduced in supabase-js 1 and 2, with gotrue-js 1.24 and 2.2.2.

## Additional context

I'm a dummy, so for all I know this is a feature request rather than a bug report, but it drove me crazy for a hot minute until I figured it out.

#### Workaround/fix:

I fixed this for myself by patching GoTrueClient's `_isImplicitGrantFlow` method to check for a custom flag in the `state` argument that I give to the OAuth provider, preventing `_getSessionFromUrl` from triggering unintentionally.

```
_isImplicitGrantFlow() {
        return (isBrowser() &&
            (Boolean(getParameterByName('access_token')) ||
                Boolean(getParameterByName('error_description')))
                && Boolean(!getParameterByName('state')?.startsWith('nogrant')));
    }
```

Another solution might involve a client option to designate a URL path where GoTrueClient will ignore `#access_token`s, and using that path in the OAuth redirect for non-login OAuth providers.

**Comment by hf** [comment]
Hey @elivox, Supabase Auth only deals with sign-in with OAuth. If you need to use OAuth in your app for other purposes (to re-authorize or ask for a bigger scope -- to read a user's contacts for example) you should implement this within your app.

Right now if you want Supabase Auth to give you back a provider token, it will be included in the callback and you are able to extract it from the `getSession()` API.

**Comment by elivox** [comment]
Hi @hf, thanks for replying! Unless I'm mistaken, your comment doesn't quite respond to the issue I've described.

If I can clarify, the issue is that as designed, Supabase Auth is _fully incompatible_ with other OAuth implementations in the same project. Having Supabase Auth in place _breaks_ non-Supabase OAuth flows, and equally, implementing non-Supabase OAuth flows _breaks_ Supabase Auth.

It's not necessary for Supabase Auth to break when combined with other OAuth flows, and indeed I've corrected that limitation for myself by patching `gotrue-js`.

If there are no plans to change this, then perhaps it should be clearly documented that Supabase Auth will not function correctly in projects using OAuth for non-login purposes?
